src/codegen/aggregation.cpp

```diff
@@ -275,83 +534,106 @@ void Aggregation::AdvanceValues(
-                              null_bitmap);
-            break;
-          }
-          default: { break; }
+        // If the aggregate is not NULL-able, avoid NULL checking altogether and
+        // generate the fast-path route.
+        if (!null_bitmap.IsNullable(aggregate_info.storage_index)) {
+          DoAdvanceValue(codegen, space, aggregate_info, update);
+        } else {
+          // else do NULL check, which will then call DoAdvanceValue if
+          // appropriate
+          DoNullCheck(codegen, space, aggregate_info, update, null_bitmap,
+                      curr_val);
         }
       }
-      agg_is_null.ElseBlock();
-      {
-        // (1)
+      agg_is_distinct.EndIf();
+
+      // A hash table lookup also produces several paths, so we need to merge
+      // the NullBitmap values
+      null_bitmap.MergeValues(agg_is_distinct, curr_val);
+    } else {
+      // Aggregation is not distinct, just advance value
+
+      // If the aggregate is not NULL-able, avoid NULL checking altogether and
+      // generate the fast-path route.
+      if (!null_bitmap.IsNullable(aggregate_info.storage_index)) {
         DoAdvanceValue(codegen, space, aggregate_info, update);
+        continue;
       }
-      agg_is_null.EndIf();
 
-      // Merge the null value
-      null_bitmap.MergeValues(agg_is_null, curr_val);
+      // do NULL check, which will call DoAdvanceValue if appropriate
+      DoNullCheck(codegen, space, aggregate_info, update, null_bitmap);
     }
-    valid_update.EndIf();
-
-    // Merge the null value
-    null_bitmap.MergeValues(valid_update, curr_val);
   }
 
   // Write the final contents of the null bitmap
   null_bitmap.WriteBack(codegen);
 }
 
+// Advance all stored aggregates (stored in the provided storage space) using
+// the values in the provided vector// Helper Function to advance global
+// aggregations
+// that creates an empty vector for the reference
+void Aggregation::AdvanceValues(CompilationContext &context, llvm::Value *space,
+                                const std::vector<codegen::Value> &next) const {
+  // Create empty vector and hand the reference to the actual implementation
+  // makes it easier to call this function without providing grouping keys
+  std::vector<codegen::Value> empty;
+  AdvanceValues(context, space, next, empty);
+}
+
 // This function will computes the final values of all aggregates stored in the
 // provided storage space, and populates the provided vector with these values.
 void Aggregation::FinalizeValues(
     CodeGen &codegen, llvm::Value *space,
     std::vector<codegen::Value> &final_vals) const {
   // Collect all final values into this map. We need this because some
   // aggregates are derived from other component aggregates.
-  std::map<std::pair<uint32_t, ExpressionType>, codegen::Value> vals;
+  // bool shows if distinct
+  std::map<std::tuple<uint32_t, ExpressionType, bool>, codegen::Value> vals;
```

Comment:
Why does the map need to know if it's a distinct aggregate?

Reply:
A distinct SUM and a non-distinct one have different results, and for a distinct AVG I have to pick the correct one.